### PR TITLE
Add support to TornadoNativeArray data types to get Memory Segments without header offset

### PR DESCRIPTION
**Pull request by mikepapadim**
#### Description

This PR adds support to obtain a MemorySegment from TornadoNativeArray types with and without the header bytes. 

* ``MemorySegment getSegment()`` returns a memory segment excluding  the Tornado header.
* ``MemorySegment getSegmentWithHeader()`` returns a memory segment as allocated for TornadoVM including the Tornado header (24bytes).
*  ``long getNumBytesOfSegmentWithHeader()`` returns Segment's number of bytes including the Tornado header.
* ``long getNumBytesOfSegment()`` returns Segment's number of bytes excluding the Tornado header.

By having distinct funtion in the above auxiliry methods we can have interoperoability when using Segments for GPU execution through the Taskgraph API and Vector API when using any of the TornadoNative types.

Also, refactor internally methods that operate on memory segments to be named explictly as ``getSegmentWithHeader``.

#### Backend/s tested

Mark the backends affected by this PR.

- [x] OpenCL
- [x] PTX
- [x] SPIRV

#### OS tested

Mark the OS where this PR is tested.

- [x] Linux
- [ ] OSx
- [ ] Windows

#### Did you check on FPGAs?

If it is applicable, check your changes on FPGAs.

- [ ] Yes
- [x] No

#### How to test the new patch?
```bash 
make jdk21  BACKEND=openc
make tests 
```
----------------------------------------------------------------------------

### CODE REVIEW

**Review by stratika — COMMENTED** [review]
I noticed that in many of the classes, the year in the copyright header has not been updated to 2024.

**Comment by mikepapadim** [comment]
> I am confused. The `withHeader` version, does it always return +24 bytes, or it returns the bytes specified in `TornadoNativeArray.ARRAY_HEADER`?

`withHeader` returns always number of bytes including the `TornadoNativeArray.ARRAY_HEADER`

File: docs/source/offheap-types.rst [hunk 1]
```diff
@@ -37,7 +37,8 @@ Additionally, developers can create an instance of a TornadoVM native array by i
 The main methods that the off-heap types expose to manage the Memory Segment of each type are presented in the list below. 
 
 .. code:: java
-
+   public MemorySegment getSegment() // returns the Memory Segment that data is stored as a slice by skipping the internal offset
```
Comment (jjfumero) [inline docs/source/offheap-types.rst:40]: what is the internal offset? I would rephrase it to 
```
Returns the memory segment without the Tornado Array Header
```

File: docs/source/offheap-types.rst [hunk 2]
```diff
@@ -37,7 +37,8 @@ Additionally, developers can create an instance of a TornadoVM native array by i
...
+   public MemorySegment getSegmentWithHeader() //  returns the Memory Segment that data is stored including the internal offset
```
Comment (jjfumero) [inline docs/source/offheap-types.rst:41]: Same here

**Comment by jjfumero** [comment]
> > I am confused. The `withHeader` version, does it always return +24 bytes, or it returns the bytes specified in `TornadoNativeArray.ARRAY_HEADER`?
> 
> `withHeader` returns always number of bytes including the `TornadoNativeArray.ARRAY_HEADER`

So please, update the PR description with this.

File: tornado-api/src/main/java/uk/ac/manchester/tornado/api/DataRange.java [hunk 3]
```diff
@@ -29,23 +29,18 @@
 public class DataRange {
 
     private final TornadoNativeArray tornadoNativeArray;
+    private final long totalSizeInBytes;
+    private final int elementSize;
```
Comment (jjfumero) [inline tornado-api/src/main/java/uk/ac/manchester/tornado/api/DataRange.java:33]: Out of curiosity, if we all use the same IntelliJ templates for code formatting, why is gets shuffled every time?

File: tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/arrays/ByteArray.java [hunk 4]
```diff
@@ -190,6 +190,16 @@ public int getSize() {
      */
     @Override
     public MemorySegment getSegment() {
+        return segment.asSlice(TornadoNativeArray.ARRAY_HEADER);
+    }
+
+    /**
+     * Returns the underlying {@link MemorySegment} of the {@code ByteArray} instance, including the header.
```
Comment (jjfumero) [inline tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/arrays/ByteArray.java:197]: Substitute all `{@code XXX}` with `{@link XXX}` in the Java doc for the new functions.

**Comment by mikepapadim** [comment]
> > > I am confused. The `withHeader` version, does it always return +24 bytes, or it returns the bytes specified in `TornadoNativeArray.ARRAY_HEADER`?
> > 
> > 
> > `withHeader` returns always number of bytes including the `TornadoNativeArray.ARRAY_HEADER`
> 
> So please, update the PR description with this.

done

File: tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/arrays/TornadoNativeArray.java [hunk 5]
```diff
@@ -49,25 +49,32 @@ public abstract sealed class TornadoNativeArray permits ByteArray, CharArray, Do
     public abstract int getSize();
 
     /**
-     * Returns the underlying {@link MemorySegment} of the native array.
+     * Returns the underlying {@link MemorySegment} of the native array, without including the header.
```
Comment (jjfumero) [inline tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/arrays/TornadoNativeArray.java:52]:
```suggestion
     * Returns the underlying {@link MemorySegment} of the native array, without the Tornado Array header.
```

**Review by jjfumero — CHANGES_REQUESTED** [review]

**Comment by jjfumero** [comment]
Unittests are passing on OSx 14.4

File: tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/arrays/ByteArray.java (hunk 4 above)
Comment (mikepapadim) [reply tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/arrays/ByteArray.java:197]: these were introduced in the original PR, but i ll fix them

**Comment by mikepapadim** [comment]
@jjfumero done

**Comment by jjfumero** [comment]
ok, thanks. the PR looks good to me.

**Review by jjfumero — APPROVED** [review]

**Review by mairooni — APPROVED** [review]
